litejpeg/core/rle/rlecore.py

```diff
@@ -0,0 +1,303 @@
+        ]
+
+        self.sync += [
+            self.dovalid_next.eq(self.dovalid),
+            self.dovalid_next_next.eq(self.dovalid_next),
+        ]
+
+        self.sync += [
+
+            # Connecting the Datapath module to the main module.
+            self.source_inter.data[0:12].eq(accumulator),
+            self.source_inter.data[12:16].eq(runlength),
+            self.source_inter.data[16].eq(self.dovalid)
+
+        ]
+
+        self.sync += [
+            self.source.data.eq(self.source_inter.data)
+        ]
+
+
+class RunLength(PipelinedActor, Module):
+    """
+    This module will connect the RLE core datapath with the input
+    and output either from other modules or from the Test Benches.
+    The input is been taken from the sink and source and is been
+    transferred to the RLE core datapath by using read and write count.
+    The RLEDatapath will than calculate the number of zeros between two
+    consecutive non-zero numbers and give the output as runlength.
+
+    Attributes :
+    ------------
+    sink : 12 bits
+           receives data from the RLEmain containing the amplitude.
+    source : 17 bits
+             transmit output to the RLEmain
+             12 bits : amplitude
+             4 bits : runlength
+             1 bit : data_valid
+    """
+    def __init__(self):
+
+        # Connecting the module to the input and the output.
+        self.sink = sink = stream.Endpoint(
+                               EndpointDescription(block_layout(12)))
+        self.source = source = stream.Endpoint(
+                                   EndpointDescription(block_layout(17)))
+
+        # Adding PipelineActor to provide additional clock for the module.
+        # This clock is useful to compensate the latency caused by the
+        # datapath to process the first input.
+        PipelinedActor.__init__(self, datapath_latency)
+        self.latency = datapath_latency
+
+        # Connecting RLE submodule.
+        self.submodules.datapath = RLEDatapath()
+        self.comb += self.datapath.ce.eq(self.pipe_ce)
+
+        BLOCK_COUNT = 64
+        # Check wheather to start write or not.
+        write_sel = Signal()
+        # To swap the write select.
+        write_swap = Signal()
+        # Check wheather to start read or not.
+        read_sel = Signal(reset=1)
+        # To swap the read_sel.
+        read_swap = Signal()
+
+        # To swap the read and write select whenever required.
+        self.sync += [
+            If(write_swap,
+               write_sel.eq(~write_sel)),
+            If(read_swap,
+               read_sel.eq(~read_sel))
+        ]
+
+        # write path
+
+        # To start the write_count back to 0.
+        write_clear = Signal()
+        # To increment the write_count.
+        write_inc = Signal()
+        # To keep track over which value of the matrix is under process.
+        write_count = Signal(6)
+
+        # For tracking the data adress.
```

Comment:
BTW For migen we normally wrap the code like this;

```python
       self.sync += \
            If(write_clear,
                write_count.eq(0)
            ).Elif(write_inc,
                write_count.eq(write_count + 1)
            )
```
IE
```python
  If(<condition>,
  <4 spaces>statement
  ).Elif(<condition>,
  <4 spaces>statement1,
  <4 spaces>statement2,
  ).Else(
  <4 spaces>statement1,
  <4 spaces>statement2,
  )
```

Reply:
DONE to most of the places.